### PR TITLE
Infer returned value of .copy() method for collections

### LINKED ISSUE #1403: dict.copy() result is not inferred as dict
### Steps to reproduce

Run this code:
```python
import astroid

code = """
x = dict()
y = x.copy()
"""

ast = astroid.parse(code)

for name in ('x', 'y'):
  node = ast.locals[name][0]
  print(name, node.inferred())
```

### Current behavior
x [<Dict.dict l.2 at 0x7fffe852e110>]
y [Uninferable]

### Expected behavior
I expect `y` to be inferred as `dict` too.

### `python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"` output
2.4.2

### PR DESCRIPTION
**Pull request by timmartin**
## Description

Where Astroid can infer that the subject of a `.copy()` method call is a built-in type that supports the copy operator, we can infer that the returned value will have the same value as whatever was inferred.

This change only adds support for `list`, `dict`, `set` and `frozenset`; in theory we can apply it to anything inferred to be of type `collections.abc.MutableSequence`, but: a) that doesn't cover `set` and `frozenset` so we'd need special cases for them anyway and b) Astroid doesn't currently seem to support inferring which types implement `MutableSequence`.

`bytearray` is another type for which we could infer `.copy()` behavior, but Astroid doesn't seem to have any explicit support for it, so adding this looks like more work.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Related Issue

Closes #1403

### CODE REVIEW

File: astroid/brain/brain_builtin_inference.py [hunk 1]
```diff
@@ -892,6 +892,26 @@ def _build_dict_with_elements(elements):
...
+def _looks_like_copy_method(node):
+    func = node.func
+    return isinstance(func, nodes.Attribute) and func.attrname == "copy"
+
+
+def _has_copy_method(node):
+    return any(
+        isinstance(node, klass)
+        for klass in [nodes.Dict, nodes.List, nodes.Set, objects.FrozenSet]
+    )
```
Comment (Pierre-Sassoulas) [inline astroid/brain/brain_builtin_inference.py:904]:
```suggestion
    return isinstance(node, (nodes.Dict, nodes.List, nodes.Set, objects.FrozenSet))
```
I think this work https://docs.python.org/3/library/functions.html#isinstance

File: astroid/brain/brain_builtin_inference.py [hunk 2]
```diff
@@ -892,6 +892,26 @@ def _build_dict_with_elements(elements):
     return _build_dict_with_elements([])
 
 
+def _looks_like_copy_method(node):
```
Comment (DanielNoord) [inline astroid/brain/brain_builtin_inference.py:895]:
```suggestion
def _looks_like_copy_method(node: nodes.Call) -> bool:
```

File: astroid/brain/brain_builtin_inference.py [hunk 3]
```diff
@@ -892,6 +892,26 @@ def _build_dict_with_elements(elements):
...
+    func = node.func
+    return isinstance(func, nodes.Attribute) and func.attrname == "copy"
+
+
+def _has_copy_method(node):
```
Comment (DanielNoord) [inline astroid/brain/brain_builtin_inference.py:900]:
```suggestion
def _has_copy_method(node: nodes.NodeNG) -> bool:
```

File: astroid/brain/brain_builtin_inference.py (hunk 1 above)
Comment (DanielNoord) [inline astroid/brain/brain_builtin_inference.py:904]:
```suggestion
    return isinstance(node, (nodes.Dict, nodes.List, nodes.Set, objects.FrozenSet))
```

File: astroid/brain/brain_builtin_inference.py [hunk 4]
```diff
@@ -892,6 +892,26 @@ def _build_dict_with_elements(elements):
...
+    return any(
+        isinstance(node, klass)
+        for klass in [nodes.Dict, nodes.List, nodes.Set, objects.FrozenSet]
+    )
+
+
+def _infer_copy_method(node, context=None):
```
Comment (DanielNoord) [inline astroid/brain/brain_builtin_inference.py:907]:
```suggestion
def _infer_copy_method(node: nodes.Call, context: Optional[InferenceContext]=None) -> ...:
```

I think the return value is `Iterator[Union[nodes.Dict, nodes.List, nodes.Set, objects.FrozenSet]]`?

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Thanks for working on this! Just some nitpicks about typing and the `isinstance` calls. Test seem good!

File: astroid/brain/brain_builtin_inference.py [hunk 5]
```diff
@@ -892,6 +892,26 @@ def _build_dict_with_elements(elements):
...
+    inferred = list(node.func.expr.infer(context=context))
```
Comment (cdce8p) [inline astroid/brain/brain_builtin_inference.py:908]: With `list` you evaluate the whole iterator even if the condition would later exit early. The work is done. [itertools.tee()](https://docs.python.org/3/library/itertools.html#itertools.tee) should work here.

File: astroid/brain/brain_builtin_inference.py [hunk 6]
```diff
@@ -892,6 +892,26 @@ def _build_dict_with_elements(elements):
...
+    if all(_has_copy_method(x) for x in inferred):
```
Comment (cdce8p) [inline astroid/brain/brain_builtin_inference.py:909]: With the suggestions `_has_copy_method` is a single `isinstance` call. We could safe multiple function lookups / calls if we just inline it here.

File: tests/unittest_inference.py [hunk 7]
```diff
@@ -2051,6 +2051,30 @@ def test_dict_invalid_args(self) -> None:
...
+        b = a.copy()
+        b #@
+
+        a = set([1, 2, 3])
+        b = a.copy()
+        b #@
+
+        a = frozenset([1, 2, 3])
+        b = a.copy()
+        b #@
```
Comment (cdce8p) [inline tests/unittest_inference.py:2070]: I would suggest to choose different variable names for each case. That could make debugging a bit easier if something were to go wrong at some point.

Could you also add a case for which we **can't** infer `copy()`?

File: ChangeLog [hunk 8]
```diff
@@ -16,6 +16,10 @@ Release date: TBA
 * Rename ``ModuleSpec`` -> ``module_type`` constructor parameter to match attribute
   name and improve typing. Use ``type`` instead.
 
+* Infer the return value of the ``.copy()`` method on ``dict``, ``list``, ``set``
```
Comment (cdce8p) [inline ChangeLog:19]:
```suggestion
* Infer the return value of the ``.copy()`` method on ``dict``, ``list``, ``set``,
```

**Review by cdce8p — CHANGES_REQUESTED** [review]

File: astroid/brain/brain_builtin_inference.py [hunk 9]
```diff
@@ -892,6 +893,21 @@ def _build_dict_with_elements(elements):
     return _build_dict_with_elements([])
 
 
+def _infer_copy_method(
+    node: nodes.Call, context: Optional[InferenceContext] = None
+) -> Iterator[Union[nodes.Dict, nodes.List, nodes.Set, objects.FrozenSet]]:
```
Comment (cdce8p) [inline astroid/brain/brain_builtin_inference.py:898]:
```suggestion
) -> Iterator[nodes.NodeNG]:
```
The result is only use the check the function itself, nothing else. So it's not necessary to constrict it too much.

File: astroid/brain/brain_builtin_inference.py [hunk 10]
```diff
@@ -4,8 +4,9 @@
 
 """Astroid hooks for various builtins."""
 
+import itertools
 from functools import partial
-from typing import Optional
+from typing import Iterator, Optional, Union
```
Comment (cdce8p) [inline astroid/brain/brain_builtin_inference.py:9]:
```suggestion
from typing import Iterator, Optional
```

File: astroid/brain/brain_builtin_inference.py [hunk 11]
```diff
@@ -892,6 +893,21 @@ def _build_dict_with_elements(elements):
     return _build_dict_with_elements([])
 
 
+def _infer_copy_method(
+    node: nodes.Call, context: Optional[InferenceContext] = None
+) -> Iterator[Union[nodes.Dict, nodes.List, nodes.Set, objects.FrozenSet]]:
+    inferred_orig, inferred_copy = itertools.tee(node.func.expr.infer(context=context))
```
Comment (cdce8p) [inline astroid/brain/brain_builtin_inference.py:899]:
```suggestion
    assert isinstance(node.func, nodes.Attribute)
    inferred_orig, inferred_copy = itertools.tee(node.func.expr.infer(context=context))
```
This is already checked in `register_transform` but mypy doesn't know about it.

**Review by cdce8p — COMMENTED** [review]
The changes look great! Just some last comments, then it's ready.

File: astroid/brain/brain_builtin_inference.py [hunk 12]
```diff
@@ -892,6 +893,22 @@ def _build_dict_with_elements(elements):
     return _build_dict_with_elements([])
 
 
+def _infer_copy_method(
+    node: nodes.Call, context: Optional[InferenceContext] = None
+) -> Iterator[nodes.nodeNG]:
```
Comment (cdce8p) [inline astroid/brain/brain_builtin_inference.py:898]:
```suggestion
) -> Iterator[nodes.NodeNG]:
```

**Review by cdce8p — APPROVED** [review]
Looks good now! Thanks @timmartin 🚀

**Pull request merged**